### PR TITLE
[CMake] Rework build and installation of all swift-syntax modules

### PR DESCRIPTION
**Pull request by DougGregor**
Rework the build and installation of all swift-syntax modules in a few ways:

* Unify the build of all of the libraries via `add_swift_host_library` to eliminate a lot of copy-and-paste.
* Respect `BUILD_SHARED_LIBS` rather than forcing static builds; the compiler will start building these as dynamic libraries
* Trick CMake into building proper `.swiftmodule` directories
* Build and install the modules into `lib/swift/host` in the toolchain, so other clients can use them

### CODE REVIEW

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please test

File: cmake/modules/AddSwiftHostLibrary.cmake [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    RUNTIME DESTINATION bin
+  )
+
+  # Install the module files.
+  install(
+    DIRECTORY ${module_base}
+    DESTINATION lib/${SWIFT_HOST_LIBRARIES_SUBDIRECTORY}
+    FILES_MATCHING PATTERN "*.swift*"
+  )
+endfunction()
```
Comment (etcwilde) [inline cmake/modules/AddSwiftHostLibrary.cmake:70]: missing file-ending newline :)

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please test

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please test Linux

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please clean test macOS

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please test

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please test

**Comment by DougGregor** [comment]
I'm turning off building with library evolution for the moment, because it's triggering some surprising failures in parsing

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please test

File: cmake/modules/AddSwiftHostLibrary.cmake [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  add_library(${name} ${ASHL_SOURCES})
+
+  # Add this to the list of exported targets.
+  set_property(GLOBAL APPEND PROPERTY SWIFTSYNTAX_EXPORTS ${name})
+
+  # Determine where Swift modules will be built and installed.
+  set(module_dir ${CMAKE_LIBRARY_OUTPUT_DIRECTORY})
+  set(module_base "${module_dir}/${name}.swiftmodule")
+  set(module_file "${module_base}/${SWIFT_MODULE_TRIPLE}.swiftmodule")
+  set(module_interface_file "${module_base}/${SWIFT_MODULE_TRIPLE}.swiftinterface")
```
Comment (ahoppen) [inline cmake/modules/AddSwiftHostLibrary.cmake:23]: I think we don’t use `module_interface_file` anywhere. Is that intentional?

Comment (DougGregor) [reply cmake/modules/AddSwiftHostLibrary.cmake:23]: It's a vestige of the failed attempt to enable library evolution. There's a potential miscompile in the very old Swift on CI.

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/62326

@swift-ci please test

**Pull request merged**